RouteReportPage.js added table pagination

kriistiina opened the pull request:
[fix] Browser slowdown after executing a Route report with large number of records; 
The 4000 row preview slice limit is removed.

Forum [Discussion](https://www.traccar.org/forums/topic/question-about-adding-pagination-for-report-visualization/)
Library used: [Material-UI table pagination](https://mui.com/material-ui/react-table/#sticky-header)

This is an initial version using the MUI example for table pagination. 
Additional features/behavior may be added after discussion and approval.

Commit:
RouteReportPage.js added table pagination

tananaev commented:
I think we discussed making it optional if there number exceeds some limit.

Can you also provide screenshots of what it looks like in desktop and mobile views.

Commit:
Add Show / Hide Functionality to RouteReportPage.js

kriistiina committed:
Merge branch 'traccar:master' into master

kriistiina commented:
Yes, with some customization, the pagination component can be shown or hidden on a condition. 
With the changes I just added, the initial views of Route report look like this for [desktop](https://github.com/traccar/traccar-web/assets/70015610/f936b731-c5a5-4ec2-b7de-35f5da189ed2) and [mobile](https://github.com/traccar/traccar-web/assets/70015610/e73bc44f-2133-4907-8e7d-b5a1ffb9d011).

After execution of Route report with large number of records, the pagination component is shown and looks like this for [desktop](https://github.com/traccar/traccar-web/assets/70015610/a6f8a03a-4c85-40ba-b283-b852ee7cbb6b) and [mobile](https://github.com/traccar/traccar-web/assets/70015610/dee80c48-c446-47f8-ae03-02c15e74d91e).

On a following report execution, the pagination component is shown or hidden, based on the number of records in the new report. Reports with less records than the first pagination step do not contain pagination and look like this for [desktop](https://github.com/traccar/traccar-web/assets/70015610/2fdbbc19-76dc-41a5-8bb6-f0c9ceb18876) and [mobile](https://github.com/traccar/traccar-web/assets/70015610/f5c5c694-ab94-4c16-9928-c49bfe5a92de).

tananaev commented on modern/src/reports/RouteReportPage.js:
```diff
@@ -164,6 +176,18 @@ const RouteReportPage = () => {
           </Table>
+          {items.length > 100 && (
+            <TablePagination
+              sx={{ position: 'sticky', bottom: 0, left: 0, backgroundColor: 'Background' }}
```
I don't like the style here. I think the pagination should be embedded in the layout instead of using "sticky". And ideally it should be in the middle of page at the bottom with proper padding. And you should use class styling.

Also, what is the "Background" color?

kriistiina replied:
I cannot seem to understand what "embedded in the layout" exactly means here. Can you please clarify?

tananaev replied:
`position` => `static`

tananaev commented on modern/src/reports/RouteReportPage.js:
```diff
@@ -164,6 +176,18 @@ const RouteReportPage = () => {
+          {items.length > 100 && (
+            <TablePagination
+              sx={{ position: 'sticky', bottom: 0, left: 0, backgroundColor: 'Background' }}
+              rowsPerPageOptions={[100, 150, 200]}
```
I think we should have something like 50, 100, 200, 500.

tananaev commented on modern/src/reports/RouteReportPage.js:
```diff
@@ -38,7 +38,19 @@ const RouteReportPage = () => {
   const [items, setItems] = useState([]);
   const [loading, setLoading] = useState(false);
   const [selectedItem, setSelectedItem] = useState(null);
+  //
```
Remove comments

tananaev commented on modern/src/reports/RouteReportPage.js:
```diff
@@ -38,7 +38,19 @@ const RouteReportPage = () => {
+  const [page, setPage] = useState(0);
+  const [rowsPerPage, setRowsPerPage] = useState(100);
 
+  const handleChangePage = (event, newPage) => {
+    setPage(newPage);
+  };
```
This can be inlined.

tananaev commented on modern/src/reports/RouteReportPage.js:
```diff
@@ -38,7 +38,19 @@ const RouteReportPage = () => {
   const [selectedItem, setSelectedItem] = useState(null);
+  //
+  const [page, setPage] = useState(0);
+  const [rowsPerPage, setRowsPerPage] = useState(100);
```
I would rename to `itemsPerPage`.

tananaev commented on modern/src/reports/RouteReportPage.js:
```diff
@@ -38,7 +38,19 @@ const RouteReportPage = () => {
+  };
+
+  const handleChangeRowsPerPage = (event) => {
+    setRowsPerPage(+event.target.value);
```
Use proper explicit conversion and only if needed. I would also inline it.

tananaev reviewed (commented):
When I said exceeding some limit, I didn't really mean one page. I was thinking something like 4k limit that we had there or at least 1-2k.

Commit:
[fix] Pagination active after 1k records;
[fix] Pagination styling moved to useStyle;
[fix] rowsPerPageOptions updated elements;
[fix] Removed comments;
[fix] onPageChange inlined;
[fix] rowsPerPage renamed to itemsPerPage;
[fix] onRowsPerPageChange inlined;

Commit:
Merge remote-tracking branch 'origin/master' into master

tananaev commented on modern/src/reports/common/useReportStyles.js:
```diff
@@ -46,4 +46,14 @@ export default makeStyles((theme) => ({
     flexGrow: 1,
     overflow: 'hidden',
   },
+  pagination: {
+    position: 'static',
+    bottom: 0,
+    '& .MuiTablePagination-spacer': {
+      display: 'none',
+    },
+    '& .MuiTablePagination-toolbar': {
+      justifyContent: 'center',
+    },
+  },
```
What is this for?

tananaev commented on modern/src/reports/RouteReportPage.js:
```diff
@@ -137,7 +140,7 @@ const RouteReportPage = () => {
             </TableHead>
             <TableBody>
-              {!loading ? items.slice(0, 4000).map((item) => (
+              {!loading ? items.slice(page * itemsPerPage, page * itemsPerPage + itemsPerPage).map((item) => (
```
Have you tested that it shows 1000 if less than 1000?

LiGuru replied:
What you think, is it tested? 

`              {!loading ? items.slice(page * itemsPerPage, items.length > 1000 ? page * itemsPerPage + itemsPerPage : items.length).map((item, id) => (
`
😁

tananaev replied:
I think it's too complicated and not very readable. Also we don't need to do `slice` if the array fits.

mohammadraufzahed replied:
Hey there,

I wanted to suggest adding virtualization, like react-window, to improve the table's performance. It can optimize rendering, especially for large datasets, by only rendering visible rows. This can enhance the user experience by reducing memory usage and providing smoother scrolling.

Consider exploring the integration of react-window for better efficiency. It's a popular library for efficient virtualization in React applications.

Thanks for considering this suggestion!

Best regards,
Mohammad Raufzahed

Commit:
[fix] show all items when less than 1000;
[fix] avoid .slice() when the array fits;

kriistiina commented:
@tananaev 
Sorry about the delay, I was on a business trip last few days. I fixed the issue regarding the visualization of all items when their count is less than 1000 and made the .slice conditional only if the items count exceeds 1000.

tananaev commented:
What do you think about the comment of using the "react-window" or something similar? The same way we do on the main screen for devices.

Commit:
[fix] Route report using react-virtuoso with Table component;
[fix] infinite scroll in Table component with always visible ReportFilter and TableHead;

kriistiina commented:
I used [react-virtuoso](https://github.com/petyosi/react-virtuoso) to load all report items in a Table component, based on this [example](https://mui.com/material-ui/react-table/#virtualized-table). This way, the behavior stays close to the one before with infinite scroll, but without the browser slowdown.

tananaev commented on modern/package.json:
```diff
@@ -66,6 +66,7 @@
     "eslint-plugin-react": "^7.32.2",
-    "eslint-plugin-react-hooks": "^4.6.0"
+    "eslint-plugin-react-hooks": "^4.6.0",
+    "react-virtuoso": "^4.3.10"
```
Why do we need this here in dev dependencies?

tananaev commented on modern/src/reports/common/useReportStyles.js:
```diff
@@ -17,6 +17,7 @@ export default makeStyles((theme) => ({
     position: 'sticky',
     left: 0,
     display: 'flex',
+    height: '100%',
```
Please explain this.

tananaev commented on modern/src/reports/RouteReportPage.js:
```diff
@@ -128,43 +184,15 @@ const RouteReportPage = () => {
-          </Table>
         </div>
+        {!loading && (
+          <TableVirtuoso
```
We already have `react-window`. Why are we adding another library that basically does the same thing?

tananaev replied:
Obviously I didn't mean that it's ok to just add more libraries that do the same thing. If you want to replace the virtualization library, you have to provide justification why the old one is bad and the new one is better. Then migrate all existing use cases to the new library.

LiGuru replied:
Ok, let's clarify.
Kristina should implement rendering with react-window.

Kristina should not use additional libraries without clarification by you.

What about adding filters and sorting?

Commit:
[fix] Route report using react-window FizedSizeGrid component;
[fix] Removed additional package react-virtuoso;
[fix] Infinite scroll with always visible ReportFilter;

kriistiina commented:
I managed to visualize the Route report by using the [FixedSizeGrid](https://react-window.vercel.app/#/api/FixedSizeGrid) component from [react-window](https://react-window.vercel.app/#/examples/grid/fixed-size) and removed the additional [react-virtuoso](https://virtuoso.dev/hello-table/).

tananaev commented on modern/src/reports/RouteReportPage.js:
```diff
@@ -128,42 +135,94 @@ const RouteReportPage = () => {
+                <FixedSizeGrid
+                  height={height}
+                  width={width}
+                  columnCount={columns.length + 2}
+                  columnWidth={(columns.length + 2) * (width * (phone ? 0.26 : desktop ? 0.1 : 0.18)) >= width ?
+                    width * (phone ? 0.26 : desktop ? 0.1 : 0.18) : width / (columns.length + 2)}
+                  rowCount={items.length > 0 ? items.length : 5}
+                  rowHeight={52}
```
This does not look right at all. Why are there a million hardcoded numbers and some crazy calculations?

tananaev commented on modern/src/reports/RouteReportPage.js:
```diff
@@ -128,42 +135,94 @@ const RouteReportPage = () => {
+                  {({ columnIndex, rowIndex, style }) => {
+                    const item = items[rowIndex];
+                    const columnKey = columns[columnIndex - 2];
+                    return (
```
The content also doesn't look good. Very hard to understand what's going on.
